File: backend/app/schemas/saved_place.py
"""Validation schemas for user-customized trip places."""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import field_validator, model_validator
from sqlmodel import Field, SQLModel

from app.schemas.place import PlaceRead


class AssignmentMode(str, Enum):
    AUTO = "AUTO"
    LOCKED = "LOCKED"


class SavedPlaceCreate(SQLModel):
    place_id: UUID
    custom_order: int | None = Field(default=None, ge=1)
    priority: int = Field(default=0, ge=0)
    is_locked: bool = False
    must_visit: bool = False
    assignment_mode: AssignmentMode = AssignmentMode.AUTO
    assigned_day_id: UUID | None = None
    notes: str | None = Field(default=None, max_length=1000)

    @field_validator("notes")
    @classmethod
    def normalize_notes(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None

    @model_validator(mode="after")
    def validate_assignment(self) -> "SavedPlaceCreate":
        if self.assignment_mode == AssignmentMode.LOCKED:
            if self.assigned_day_id is None:
                raise ValueError("assigned_day_id is required when assignment_mode is LOCKED.")
        elif self.assignment_mode == AssignmentMode.AUTO:
            if self.assigned_day_id is not None:
                raise ValueError("assigned_day_id must be null when assignment_mode is AUTO.")
        return self


class SavedPlaceUpdate(SQLModel):
    notes: str | None = Field(default=None, max_length=1000)
    priority: int | None = Field(default=None, ge=0)
    is_locked: bool | None = None
    must_visit: bool | None = None
    custom_order: int | None = Field(default=None, ge=1)
    assignment_mode: AssignmentMode | None = None
    assigned_day_id: UUID | None = None

    @field_validator("notes")
    @classmethod
    def normalize_notes(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None

    @model_validator(mode="after")
    def validate_assignment(self) -> "SavedPlaceUpdate":
        if self.assignment_mode == AssignmentMode.LOCKED:
            if self.assigned_day_id is None:
                raise ValueError("assigned_day_id is required when assignment_mode is LOCKED.")
        elif self.assignment_mode == AssignmentMode.AUTO:
            if self.assigned_day_id is not None:
                raise ValueError("assigned_day_id must be null when assignment_mode is AUTO.")
        return self


class SavedPlaceOrder(SQLModel):
    place_id: UUID
    custom_order: int = Field(ge=1)


class SavedPlaceReorder(SQLModel):
    places: list[SavedPlaceOrder] = Field(min_length=1, max_length=500)

    @model_validator(mode="after")
    def validate_unique_contiguous_order(self) -> "SavedPlaceReorder":
        place_ids = [item.place_id for item in self.places]
        orders = [item.custom_order for item in self.places]
        if len(set(place_ids)) != len(place_ids):
            raise ValueError("Each saved place may appear only once.")
        if len(set(orders)) != len(orders):
            raise ValueError("Each custom order value must be unique.")
        if sorted(orders) != list(range(1, len(orders) + 1)):
            raise ValueError("Custom order values must be contiguous from 1.")
        return self


class SavedPlaceRead(SQLModel):
    id: UUID
    trip_id: UUID
    place_id: UUID
    custom_order: int
    priority: int
    is_locked: bool
    must_visit: bool
    assignment_mode: AssignmentMode = AssignmentMode.AUTO
    assigned_day_id: UUID | None = None
    notes: str | None
    created_at: datetime
    place: PlaceRead

File: lib/models/saved_place.dart
import 'place.dart';

enum AssignmentMode {
  auto('AUTO'),
  locked('LOCKED');

  const AssignmentMode(this.value);

  final String value;

  static AssignmentMode fromJson(String? value) {
    if (value == null) return AssignmentMode.auto;
    switch (value.toUpperCase()) {
      case 'LOCKED':
        return AssignmentMode.locked;
      case 'AUTO':
      default:
        return AssignmentMode.auto;
    }
  }

  String toJson() => value;
}

class SavedPlace {
  const SavedPlace({
    required this.id,
    required this.tripId,
    required this.placeId,
    required this.customOrder,
    required this.priority,
    required this.isLocked,
    required this.mustVisit,
    required this.place,
    this.assignmentMode = AssignmentMode.auto,
    this.assignedDayId,
    this.notes,
  });

  final String id;
  final String tripId;
  final String placeId;
  final int customOrder;
  final int priority;
  final bool isLocked;
  final bool mustVisit;
  final AssignmentMode assignmentMode;
  final String? assignedDayId;
  final String? notes;
  final Place place;

  factory SavedPlace.fromJson(Map<String, dynamic> json) {
    return SavedPlace(
      id: json['id'] as String,
      tripId: json['trip_id'] as String,
      placeId: json['place_id'] as String,
      customOrder: json['custom_order'] as int,
      priority: json['priority'] as int? ?? 0,
      isLocked: json['is_locked'] as bool? ?? false,
      mustVisit: json['must_visit'] as bool? ?? false,
      assignmentMode: AssignmentMode.fromJson(
        json['assignment_mode'] as String?,
      ),
      assignedDayId: json['assigned_day_id'] as String?,
      notes: json['notes'] as String?,
      place: Place.fromJson(json['place'] as Map<String, dynamic>),
    );
  }

  SavedPlace copyWith({
    String? id,
    String? tripId,
    String? placeId,
    int? customOrder,
    int? priority,
    bool? isLocked,
    bool? mustVisit,
    AssignmentMode? assignmentMode,
    String? assignedDayId,
    bool clearAssignedDay = false,
    String? notes,
    Place? place,
  }) {
    return SavedPlace(
      id: id ?? this.id,
      tripId: tripId ?? this.tripId,
      placeId: placeId ?? this.placeId,
      customOrder: customOrder ?? this.customOrder,
      priority: priority ?? this.priority,
      isLocked: isLocked ?? this.isLocked,
      mustVisit: mustVisit ?? this.mustVisit,
      assignmentMode: assignmentMode ?? this.assignmentMode,
      assignedDayId:
          clearAssignedDay ? null : (assignedDayId ?? this.assignedDayId),
      notes: notes ?? this.notes,
      place: place ?? this.place,
    );
  }
}
